Added Elevator subsystem

MiscellaneousShape opened the pull request.

MiscellaneousShape committed:
Added Elevator subsystem

ShamuelBoy reviewed (approved):
Ayo looks good to me bub

bradleyscollins commented on src/main/cpp/subsystems/Elevator.cpp:
Indentation in this file is inconsistent with the rest of the project. Everywhere else we use 2 spaces for indentation.

bradleyscollins commented on src/main/cpp/subsystems/Elevator.cpp:
```diff
@@ -0,0 +1,92 @@
+#include "subsystems/Elevator.h"
+
+Elevator::Elevator()
+//: m_motors(elevator_motor_id)
```
Please remove this stray comment

bradleyscollins commented on src/main/include/subsystems/Elevator.h:
```diff
@@ -0,0 +1,40 @@
+  SparkMax m_wrist_motor;
+
+  VictorSPX m_vert_motor_left;
+  VictorSPX m_vert_motor_right;
```
Suggest avoiding abbreviations:

```suggestion
  VictorSPX m_vertical_motor_left;
  VictorSPX m_vertical_motor_right;
```

bradleyscollins commented on src/main/include/subsystems/Elevator.h:
```diff
@@ -0,0 +1,40 @@
+  VictorSPX m_vert_motor_right;
+
+  frc::PIDController m_vert_motors_PID;
+  frc::PIDController m_wrist_motor_PID;
```
Naming convention (and a correction):

```suggestion
  frc::PIDController m_vertical_motor_pid;
  frc::PIDController m_wrist_motor_pid;
```

bradleyscollins commented on src/main/include/subsystems/Elevator.h:
```diff
@@ -0,0 +1,40 @@
+#include <rev/SparkMax.h>
+#include <frc2/command/CommandScheduler.h>
+#include <ctre/phoenix/motorcontrol/can/VictorSPX.h>
+#include "Constants.h"
```
Not the end of the world, but it's nice when includes are organized & alphabetized—especially when there are more than just a few—so that it's easy to see what's there and what might be missing.

```suggestion
#include <ctre/phoenix/motorcontrol/can/TalonSRX.h>
#include <ctre/phoenix/motorcontrol/can/VictorSPX.h>
#include <frc/controller/PIDController.h>
#include <frc2/command/CommandPtr.h>
#include <frc2/command/CommandScheduler.h>
#include <frc2/command/SubsystemBase.h>
#include <rev/SparkMax.h>
#include <units/length.h>

#include "Constants.h"
```

Also, as above, I usually put all my third-party includes (the stuff in `<>`'s) first and my stuff (the things in `""`'s) second.

bradleyscollins commented on src/main/cpp/subsystems/Elevator.cpp:
```diff
@@ -0,0 +1,92 @@
+
+    }).Until([this]{
+        return m_wrist_motor_PID.AtSetpoint();
+
```
Please remove this stray blank line

bradleyscollins commented on src/main/cpp/subsystems/Elevator.cpp:
```diff
@@ -0,0 +1,92 @@
+    [this]{
+        //m_motors.Set(ctre::phoenix::motorcontrol::ControlMode::PercentOutput, 0.0);
+        m_wrist_motor.StopMotor();
+
```
Stray blank line

bradleyscollins commented on src/main/cpp/subsystems/Elevator.cpp:
```diff
@@ -0,0 +1,92 @@
+
+frc2::CommandPtr Elevator::ElevateCommand(units::inch_t height)
+{
+
```
Stray blank line

bradleyscollins commented on src/main/cpp/subsystems/Elevator.cpp:
```diff
@@ -0,0 +1,92 @@
+frc2::CommandPtr Elevator::MoveWristToCommand(units::degree_t degree)
+{
+    return this->RunEnd([this, degree] {
+        //m_motors_PID.Calculate(m_motors.GetSelectedSensorPosition(), height.value());
```
Stray comment

bradleyscollins commented on src/main/cpp/subsystems/Elevator.cpp:
```diff
@@ -0,0 +1,92 @@
+            NEOUnitToDegree(m_wrist_motor.GetEncoder().GetPosition()), NEOUnitToDegree(degree.value())));
+    },
+    [this]{
+        //m_motors.Set(ctre::phoenix::motorcontrol::ControlMode::PercentOutput, 0.0);
```
Stray comment

bradleyscollins commented on src/main/cpp/subsystems/Elevator.cpp:
```diff
@@ -0,0 +1,92 @@
+{
+
+    return this->RunEnd([this, height] {
+        //m_motors_PID.Calculate(m_motors.GetSelectedSensorPosition(), height.value());
```
Stray comment

bradleyscollins commented:
We appear to be missing controller mappings. Per @ShamuelBoy, they are as follows:

- D-pad up: Move elevator up one level
- D-pad down: Move elevator down one level

bradleyscollins commented on src/main/cpp/subsystems/Elevator.cpp:
```diff
@@ -0,0 +1,92 @@
+frc2::CommandPtr Elevator::MoveUpOnceCommand() {
+    return this->RunOnce([this]{
+    level = (level < 4 ) ? level++ : level;
+    });
```
This command (and `MoveDownCommand`) only increments the level, but it does not actually move the elevator. 

Suggest that we modify `MoveToLevelCommand` to take a `level` parameter and delegate to it:

```suggestion
  return this->MoveToLevelCommand(m_level + 1); 
```

bradleyscollins commented on src/main/include/subsystems/Elevator.h:
```diff
@@ -0,0 +1,40 @@
+  Elevator();
+
+  frc2::CommandPtr ElevateCommand(units::inch_t height);
+  frc2::CommandPtr MoveWristToCommand(units::degree_t degree);
```
It really is redundant here to name the parameter the same as the type. Suggest something that is a _little_ more meaningful int the context:

```suggestion
  frc2::CommandPtr MoveWristToCommand(units::degree_t angle);
```

bradleyscollins commented on src/main/cpp/subsystems/Elevator.cpp:
```diff
@@ -0,0 +1,92 @@
+    });
+}
+
+frc2::CommandPtr Elevator::MoveToLevelCommand()
+{
+    switch (level)
+    {
+    case 0:
+        return MoveWristToCommand(units::degree_t(85))
+        .AndThen(ElevateCommand(units::inch_t(0)));
+        break;
+
+    case 1:
+        return MoveWristToCommand(units::degree_t(35))
+        .AndThen(ElevateCommand(units::inch_t(18 - BASE_HEIGHT_FROM_FLOOR_INCHES.value())));
+        break;
+        
+    case 2:
+        return MoveWristToCommand(units::degree_t(35))
+        .AndThen(ElevateCommand(units::inch_t(31 + (7/8) - BASE_HEIGHT_FROM_FLOOR_INCHES.value())));
+        break;
+        
+    case 3:
+        return MoveWristToCommand(units::degree_t(35))
+        .AndThen(ElevateCommand(units::inch_t(47 + (7/8) - BASE_HEIGHT_FROM_FLOOR_INCHES.value())));
+        break;
+        
+    case 4:
+        return MoveWristToCommand(units::degree_t(88))
+        .AndThen(ElevateCommand(units::inch_t(72 - BASE_HEIGHT_FROM_FLOOR_INCHES.value())));
+        break;
+        
+    }
+}
```
Several things here. 

First, per another comment, we should probably change this function to take a `level` parameter. You'll want to make sure that `level` is within the allowable range (0–4). If it is within range, then you should save `m_level` and build & return the command.

Second, there's a lot of duplicate code here. The only differences in each case of the `switch` are ...

- The angle, i.e., the input to `MoveWristCommand` and
- The height, i.e., the input to `ElevateCommand`.

Therefore, an approach we could take is to do nothing in the `switch` but set some local variables, `angle` and `height`, and just build the commands once.

Third, there is a better way than using a switch. You can put all your cases in a `std::array` (and take advantage of unit literals):

```cpp
static const std::array<std::pair<units::degree_t, units::inch_t>, 5> presets{
  { 85_deg, 0_in },
  { 35_deg, units::inch_t(18 - BASE_HEIGHT_FROM_FLOOR_INCHES.value()) },
  { 35_deg, units::inch_t(31 + (7/8) - BASE_HEIGHT_FROM_FLOOR_INCHES.value()) },
  { 35_deg, units::inch_t(47 + (7/8) - BASE_HEIGHT_FROM_FLOOR_INCHES.value()) },
  { 88_deg, units::inch_t(72- BASE_HEIGHT_FROM_FLOOR_INCHES.value()) },
};
```

Then you could implement this function something like this:

```suggestion
frc2::CommandPtr Elevator::MoveToLevelCommand(int level)
{
  // Make sure level is within range
  m_level = std::clamp(level, 0, static_cast<int>(presets.size() - 1));
  
  // Fetch the angle & height for the preset associated with this level
  auto [angle, height] = presets.at(m_level);

  // Build and return composite command
  return MoveWristToCommand(angle)AndThen(ElevateCommand(height));
}
```

Fourth, this `31 + (7/8)` does not do what you think. It evaluates to `31`. To get the `7/8` to evaluate to its decimal equivalent, you'll need to use floating point values.

Finally, instead of fetching the value of `BASE_HEIGHT_FROM_FLOOR_INCHES`, you can add/subtract it directly to/from another `inch_t` value:

```cpp
18_in - BASE_HEIGHT_FROM_FLOOR_INCHES
```

bradleyscollins commented on src/main/include/subsystems/Elevator.h:
```diff
@@ -0,0 +1,40 @@
+  frc2::CommandPtr MoveToLevelCommand();
+
+private:
+  int level;
```
Naming convention:

```suggestion
  int m_level;
```

bradleyscollins reviewed (changes requested):
Suggesting some improvements.

EGarcia354 commented on src/main/cpp/subsystems/Elevator.cpp:
```diff
@@ -0,0 +1,92 @@
+, m_wrist_motor_PID(0.5, 0.0, 0.0)
+, level(0)
+{
+    m_vert_motors_PID.SetTolerance(NEOUnitToInch(0.5));
```
What role does calling NeounitToinch play in the logic of this line

bradleyscollins replied:
Per @LoganR-FRC-Team34, the function is misnamed. It should be renamed to indicate that it takes a value in inches and converts it to the Neo's units. The math it performs is correct; the name is just backward.

We will perform this change in a separate pull request.

MiscellaneousShape committed:
addressed pull request comments

MiscellaneousShape committed:
addressed pull request comments

MiscellaneousShape committed:
addressed pull request comments

MiscellaneousShape committed:
addressed pull request comments

bradleyscollins committed:
Formatting only; no functional change

bradleyscollins committed:
Formatting only; no functional change

bradleyscollins committed:
Formatting only; no functional change

MiscellaneousShape committed:
addressed pull request comments

bradleyscollins committed:
refactor: Renamed ElevateCommand() to ElevateToCommand()

A member of the team mistook the intent of the command for moving relative distances rather than moving to a specific distance.

bradleyscollins committed:
Formatting only; no functional change

bradleyscollins committed:
Fixing warning messages

bradleyscollins committed:
Formatting only; no functional change

bradleyscollins committed:
Moved Elevator into t34 namespace
